Add serialization support for register_constant

tugsbayasgalan opened the pull request:
Differential Revision: D75767873




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

tugsbayasgalan commented:
@pytorchbot label "topic: not user facing"

angelayi commented on torch/utils/_pytree.py:
```diff
@@ -384,7 +384,9 @@ def _flatten_fn_with_keys(obj: Any) -> tuple[list[Any], Context]:
 
-def register_constant(cls: type[Any]) -> None:
+def register_constant(
+    cls: type[Any], serialized_type_name: Optional[str] = None
```
can we make serialized_type_name a kwarg?

angelayi commented on torch/utils/_pytree.py:
```diff
@@ -840,6 +858,17 @@ def _namedtuple_deserialize(dumpable_context: DumpableContext) -> Context:
+            "because we couldn't find a serializated name."
+        )
+
+    typ = SERIALIZED_TYPE_TO_PYTHON_TYPE[dumpable_context]
```
Hmm this doesn't seem entirely right. The treespec of just the register constant looks like: `TreeSpec(Bar, ConstantNode(value=<__main__.TestMoo.test_aoti_pytree_constant.<locals>.Bar object at 0x7f96dcbc9150>), [])`. 

As you can see, the context is a `ConstantNode` with the `Bar` object directly. 

However with this implementation of serialize/deserialize, we only get the class name, `Bar`. So the deserialized treespec now only contains the class type, instead of the object instance: `TreeSpec(Bar, <class '__main__.TestMoo.test_aoti_pytree_constant.<locals>.Bar'>, [])`.

A better test would be something like:
```
torch.utils._pytree.register_constant(Bar, serialized_type_name="Bar")
spec = pytree.tree_flatten(Bar())[1]
print(spec)
spec_new = pytree.treespec_loads(pytree.treespec_dumps(spec))
print(spec_new)
# compare spec and spec_new
```

I guess in the context of AOTI and export, how this object is serialized doesn't matter that much since the input never gets used and the values are embedded in the graph, but we should probably still implement a lossless serialization mechanism? 

A possible solution is that to_dumpable_context returns a pickle.dump-ed constant, and from_dumpable_context pickle.loads it. 

cc @zou3519 if you have any thoughts?

tugsbayasgalan replied:
updated!

zou3519 replied:
Could one of you remind me how serialized_type_name works?
"A keyword argument used to specify the fully qualified name used when serializing the tree spec. This is only needed for serializing the treespec in torch.export." doesn't tell me much

angelayi replied:
When we serialize the treespec type, we look up this serialized_type_name to serialize the type to, and when we deserialize we then look up the serialized_type_name to know which python class type to deserialize it back into. This was considered better than if we just automatically set the serialized_type_name to be `f"{type.__module__}{type.__name__}"` because if someone moved the location of the class then this serialization mechanism would be broken. So instead we force users to provide the FQN for this class.

zou3519 replied:
If we're saying the way to serialize a constant is to pickle dump and load it, do we still need the serialized_type_name?

tugsbayasgalan committed:
Add serialization support for register_constant (#154834)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/154834

Test Plan: CI

Differential Revision: D75767873
